### PR TITLE
fix(icons): add support for font awesome svg icons

### LINKED ISSUE #8016: [Feature Request] Support for faSvg in 2.0
### Problem to solve
Allows the use of Fontawesome SVG icons

### Proposed solution
This was implemented in 1.5 but was removed in 2.0 not sure why, maybe we could copy over the code from 1.5 over to 2.0? I could help with a PR if you approve.

### PR DESCRIPTION
**Pull request by basselAhmed**
## Description

Just add faSvg option to iconfont like described in the [docs.](https://vuetifyjs.com/en/customization/icons#install-font-awesome-svg-icons)

## Motivation and Context

Give the user the ability to use Font Awesome icons as SVG using Font Awesome Vue component.
resolves #8016 

## How Has This Been Tested?

Through the playground, add font awesome packages and set **iconfont** to **faSvg**.
Then you can check if the icons render correctly using <font-awesome-icon> component.
I checked for v-autocomplete dropdown icon and v-stepper-step complete icon.

## Markup:

<details>

```vue
<template>
  <v-container style="max-width: 500px">
    <!--  -->
    <v-autocomplete/>

    <v-stepper>
      <v-stepper-header>
      <v-stepper-step :complete="true" step="1">Name of step 1</v-stepper-step>

      <v-divider></v-divider>

      <v-stepper-step :complete="true" step="2">Name of step 2</v-stepper-step>

      <v-divider></v-divider>

      <v-stepper-step step="3">Name of step 3</v-stepper-step>
    </v-stepper-header>
    </v-stepper>
  </v-container>
</template>

<script>
export default {
  data: () => ({
    //
  })
}
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library) (Already there in the docs)
- [ ] I've added new examples to the kitchen (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by morphatic** [comment]
Yay! Thanks for doing this. I was just in the process of upgrading an app to Vuetify v2.x that uses FontAwesome Pro. Didn't realize that `faSvg` hadn't been implemented, yet.

I have some questions:

1. Does this implementation support tree-shaking or would I end up having to include the entire `fas` icon set into my project?
2. Would it be possible to make the name of the `font-awesome-icon` component customizable? To save on typing, I typically use `fa-icon` as the name of this component in my implementations. I don't mind switching to what the docs say, but this would be a nice feature.
3. Does this implementation still support customization? i.e. if I wanted to use a different icon for one of the `$vuetify.icons`, could I still do something like:

```js
// src/plugins/vuetify.js

export default new Vuetify({
  icons: {
    iconfont: 'faSvg',
    values: {
      next: 'fas fa-arrow-right',
      prev: 'fas fa-arrow-left'
    }
  }
}
```

**Comment by morphatic** [comment]
On the subject of tree-shaking, [the `mdiSvg` preset actually just includes the raw `svg` code](https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/services/icons/presets/mdi-svg.ts) for the standard icon set. Should we be doing the same for `faSvg`?

**Comment by basselAhmed** [comment]
@morphatic 
1. This implementation doesn't include font awesome packages, you'll have to import those yourself like the snippet below
```js
import Vue from 'vue'
import { library } from '@fortawesome/fontawesome-svg-core'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import { faCheck, faCaretDown, faCaretRight } from '@fortawesome/free-solid-svg-icons'

Vue.component('font-awesome-icon', FontAwesomeIcon)
library.add(faCheck, faCaretDown, faCaretRight)
```
This way you can only include the icons you want, but you'll need to include some specific icons used in Vuetify components like v-autocomplete uses faCaretDown and v-stepper-step uses faCheck.
complete list of those is here [list](https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/services/icons/presets/fa.ts)

2. You can set another component with a shorter name if you want, see below. I don't think Vuetify should give such an option.
```js
Vue.component('font-awesome-icon', FontAwesomeIcon)
Vue.component('fa-icon', FontAwesomeIcon) 
```
3. It supports customization but in a different format to get an SVG icon, see snippet.
```js
const vuetify = new Vuetify({
  icons: {
    iconfont: 'faSvg',
    values: {
      next: {
        component: 'font-awesome-icon',
        props: {
          icon: ['fas', 'caret-right'],
        },
      },
    },
  },
})
```
4. I thought about doing it that way, but I thought it's better to use the official font awesome component to render the icons instead of hard coding it.

**Comment by MajesticPotatoe** [comment]
@morphatic, as @basselAhmed This is indeed more of a re-enabling of the use of faSvg in 2.0. (which got somehow lost in conversion)

**Comment by SirLamer** [comment]
For what it's worth, I'd appreciate being able to define the icon with the array format right in the component. I know this is a deviation from the current standard but, for example, there are sometimes good reasons to include many of the "pro" icons throughout the application and it would be a lot simpler and clearer to implement them right in the component. This is also more similar to how vue-fontawesome works.

Example implementation:

```html
<v-icon :icon="['fas', 'caret-right']"></v-icon>
```

**Comment by SaneMethod** [comment]
Sorry to be a bug, but would be nice to get this merged in soonish - 2.x vuetify projects on our side of things are currently forced to monkey-patch this in. :(

**Comment by SaneMethod** [comment]
(For those who might come to this thread looking for a solution while waiting for this patch):

```js
import Vue from 'vue';
import Vuetify from 'vuetify/lib';

import {library as faLib} from '@fortawesome/fontawesome-svg-core';
import {FontAwesomeIcon} from '@fortawesome/vue-fontawesome';
import {fas} from '@fortawesome/free-solid-svg-icons';

// Register font awesome components for use in Vue
// See: https://github.com/FortAwesome/vue-fontawesome#usage
Vue.component('font-awesome-icon', FontAwesomeIcon);
// Add Font Awesome Solid icon font to library for vuetify support.
faLib.add(fas);

// Extend Vue with Vuetify
// See: https://vuetifyjs.com/en/components/api-explorer
Vue.use(Vuetify);

const FAIcons = {
    complete: 'fas fa-check',
    cancel: 'fas fa-times-circle',
    close: 'fas fa-times',
    delete: 'fas fa-times-circle',
    clear: 'fas fa-times-circle',
    success: 'fas fa-check-circle',
    info: 'fas fa-info-circle',
    warning: 'fas fa-exclamation',
    error: 'fas fa-exclamation-triangle',
    prev: 'fas fa-chevron-left',
    next: 'fas fa-chevron-right',
    checkboxOn: 'fas fa-check-square',
    checkboxOff: 'far fa-square',
    checkboxIndeterminate: 'fas fa-minus-square',
    delimiter: 'fas fa-circle',
    sort: 'fas fa-sort-up',
    expand: 'fas fa-chevron-down',
    menu: 'fas fa-bars',
    subgroup: 'fas fa-caret-down',
    dropdown: 'fas fa-caret-down',
    radioOn: 'far fa-dot-circle',
    radioOff: 'far fa-circle',
    edit: 'fas fa-edit',
    ratingEmpty: 'far fa-star',
    ratingFull: 'fas fa-star',
    ratingHalf: 'fas fa-star-half',
    loading: 'fas fa-sync',
    first: 'fas fa-step-backward',
    last: 'fas fa-step-forward',
    unfold: 'fas fa-arrows-alt-v',
    file: 'fas fa-paperclip',
    plus: 'fas fa-plus',
    minus: 'fas fa-minus',
};

export default new Vuetify({
    icons:{
        iconfont: 'faSvg', // Set Vuetify to use font-awesome 5 svg icon fonts
        values:(() => {
            // Patch in the fa-svg icons until https://github.com/vuetifyjs/vuetify/pull/8092 is merged
            Object.keys(FAIcons).forEach((icon) => {
                FAIcons[icon] = {
                    component:'font-awesome-icon',
                    props:{
                        icon: FAIcons[icon].split(' fa-'),
                    },
                };
            });

            return FAIcons;
        })(),
    },
});
```

**Comment by SirLamer** [comment]
@SaneMethod 

I have implemented the solution as you have proposed, it works in most respect but I am getting icon size and alignment errors in general. Are you finding this as well?

Why initialize the icons as strings and then convert them to arrays? 'fas' is default and then you can just init the 'far' ones as arrays. Faster load times.

Also forEach is slower than most other for loop solutions. Normally that doesn't matter much but since this is page loading stuff it's useful to be slick.

**Comment by SirLamer** [comment]
Here is how I modified this snippet, note it doesn't fix the positioning/size issues but it's a bit leaner to execute and cleaner to read.

```js
const FAIcons = {
  complete: 'check',
  cancel: 'times-circle',
  close: 'times',
  delete: 'times-circle',
  clear: 'times-circle',
  success: 'check-circle',
  info: 'info-circle',
  warning: 'exclamation',
  error: 'exclamation-triangle',
  prev: 'chevron-left',
  next: 'chevron-right',
  checkboxOn: 'check-square',
  checkboxOff: ['far', 'square'],
  checkboxIndeterminate: 'minus-square',
  delimiter: 'circle',
  sort: 'sort-up',
  expand: 'chevron-down',
  menu: 'bars',
  subgroup: 'caret-down',
  dropdown: 'caret-down',
  radioOn: ['far', 'dot-circle'],
  radioOff: ['far', 'circle'],
  edit: 'edit',
  ratingEmpty: ['far', 'star'],
  ratingFull: 'star',
  ratingHalf: 'star-half',
  loading: 'sync',
  first: 'step-backward',
  last: 'step-forward',
  unfold: 'arrows-alt-v',
  file: 'paperclip',
  plus: 'plus',
  minus: 'minus',
},
  FAIconsParsed = (() => {
    // Patch in the fa-svg icons until https://github.com/vuetifyjs/vuetify/pull/8092 is merged
    const icons = Object.entries(FAIcons),
      component = 'font-awesome-icon',
      payload = {};

    for (const [key, icon] of icons) {
      payload[key] = {
        component,
        props: {
          icon,
        },
      };
    }

    return payload;
  })();

export default new Vuetify({
  icons:{
    iconfont: 'faSvg', // Set Vuetify to use font-awesome 5 svg icon fonts
    values: FAIconsParsed,
  },
});
```

**Comment by SaneMethod** [comment]
Can't say I've noticed any size or alignment issues.

No claim to optimization can or shall be made for the rinky-dink placeholder solution I posted; your revision looks fine to me.

**Comment by jacekkarczmarczyk** [comment]
@SirLamer can you show the example of the alignment issue (code and screenshot)?

**Comment by nmsmith22389** [comment]
I hope this goes through soon! I've been waiting for this since v2.0 came out! 😭

File: packages/vuetify/src/services/icons/presets/fa-svg.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { VuetifyIcon, VuetifyIcons } from 'vuetify/types/services/icons'
+import { Component } from 'vue'
+
+const faIcons: VuetifyIcons = {
```
Comment (sh7dm) [inline packages/vuetify/src/services/icons/presets/fa-svg.ts]: this could be imported from `./fa` I think

**Comment by SirLamer** [comment]
> @SirLamer can you show the example of the alignment issue (codepen and screenshot)?

Sorry I didn't get to this in a timely manner, your questioning me caused me to consider that maybe I affected the problem with my own implementation.

File: packages/vuetify/types/services/icons.d.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ export interface VuetifyIconOptions {
    * fa: <a href="https://fontawesome.com/get-started/web-fonts-with-css">FontAwesome 5</a>
    * fa4: <a href="">FontAwesome 4</a> TODO: link
```
Comment (sh7dm) [inline packages/vuetify/types/services/icons.d.ts]: please add a link like on L10 here for `faSvg`

**Comment by n4ks** [comment]
It is not known when this problem will be fixed? And are there any workarounds or is it easier to use mdiSvg so far? Thanks

**Comment by Tronix117** [comment]
To full support FontAwesome, we should also be able to do:

```js
<template>
  <v-icon>{{ faBars }}</v-icon>
</template>

<script>
  import { faBars } from '@fortawesome/pro-light-svg-icons'

  export default {
    data: () => ({
      faBars
    }),
  }
</script>
```

or with TSX:
```tsx
import { faBars } from '@fortawesome/pro-light-svg-icons'

@Component
export default class MyComponent extends Vue {
  render() {
    return <v-icon>{ faBars }</v-icon>
  }
}
```

Like in the doc for material icons.
It gets us instant tree shaking and intellisense when using typescript

---
We could use SVG path from the import, however SVG Viewbox for the path is not customizable and uses the 24x24 viewbox for materialized, whereas font awesome uses 512x512 (see: https://github.com/vuetifyjs/vuetify/blob/de7af3fcf5528706f50770de7dd1b32a0f14bc8e/packages/vuetify/src/components/VIcon/VIcon.ts#L147)

**Comment by P4sca1** [comment]
Would be great if vuetify (vuetify-loader?) could automatically detect which icons are being used and import only those from fontawesome.

**Comment by johnleider** [comment]
Requesting final comments @vuetifyjs/core-contributors @vuetifyjs/core-team

File: packages/vuetify/types/services/icons.d.ts [hunk 3]
```diff
@@ -9,8 +9,9 @@ export interface VuetifyIconOptions {
    * mdi: <a href="https://materialdesignicons.com">MDI</a>
    * fa: <a href="https://fontawesome.com/get-started/web-fonts-with-css">FontAwesome 5</a>
    * fa4: <a href="">FontAwesome 4</a> TODO: link
+   * faSvg: <a href="https://fontawesome.com/how-to-use/on-the-web/using-with/vuejs">FontAwesome SVG</a>
    */
-  iconfont: 'mdi' | 'mdiSvg' | 'md' | 'fa' | 'fa4' // TODO: camelCase
+  iconfont: 'mdi' | 'mdiSvg' | 'md' | 'fa' | 'fa4' | 'faSvg' // TODO: camelCase
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/types/services/icons.d.ts]: team: this TODO doesn't seem to be valid any longer

Comment (KaelWD) [reply packages/vuetify/types/services/icons.d.ts]: I think it was for `iconfont` -> `iconFont`

Comment (albernhagen) [reply packages/vuetify/types/services/icons.d.ts]: Appreciate your work on this!  I created a PR against your branch to address this change. https://github.com/basselAhmed/vuetify/pull/1

Comment (basselAhmed) [reply packages/vuetify/types/services/icons.d.ts]: Thanks, @albernhagen merged it.

**Comment by P4sca1** [comment]
@basselAhmed Would be great if you could change the `IconFont` type in `types/services/icons.d.ts` to reflect the new `faSvg` support like so:
`export type Iconfont = 'mdi' | 'mdiSvg' | 'md' | 'fa' | 'fa4' | 'faSvg'`.
(Creating a reproduction repo to create an issue just for this small change takes too much time).

**Pull request merged**